lib/Cake/Model/Behavior/TreeBehavior.php

```diff
@@ -442,27 +442,61 @@ public function generateTreeList(Model $Model, $conditions = null, $keyPath = nu
 			$fields = array($Model->primaryKey, $Model->displayField, $left, $right);
 		}
 
-		if (!$keyPath) {
-			$keyPath = '{n}.' . $Model->alias . '.' . $Model->primaryKey;
+		$conditions = (array)$conditions;
+		if ($scope) {
+			$conditions[] = $scope;
 		}
 
-		if (!$valuePath) {
-			$valuePath = array('%s%s', '{n}.tree_prefix', '{n}.' . $Model->alias . '.' . $Model->displayField);
+		$order = $Model->escapeField($left) . ' asc';
+		$results = $Model->find('all', compact('conditions', 'fields', 'order', 'recursive'));
 
-		} elseif (is_string($valuePath)) {
-			$valuePath = array('%s%s', '{n}.tree_prefix', $valuePath);
+		return $this->formatTreeList($Model, $results, compact('keyPath', 'valuePath', 'spacer'));
+	}
 
-		} else {
-			array_unshift($valuePath, '%s' . $valuePath[0], '{n}.tree_prefix');
+/**
+ * Formats result of a find() call to a hierarchical array used for HTML select boxes.
+ *
+ * Note that when using your own find() call this expects the order to be "left" field asc in order
+ * to generate the same result as using generateTreeList() directly.
+ *
+ * Options:
+ *
+ * - 'keyPath': A string path to the key, i.e. "{n}.Post.id"
+ * - 'valuePath': A string path to the value, i.e. "{n}.Post.title"
+ * - 'spacer': The character or characters which will be repeated
+ *
+ * @param Model $Model Model using this behavior
+ * @param array $results Result array of a find() call
+ * @param array $options Options
+ * @return array An associative array of records, where the id is the key, and the display field is the value
+ */
+	public function formatTreeList(Model $Model, array $results, array $options = array()) {
+		if (empty($results)) {
+			return array();
 		}
+		$defaults = array(
+			'keyPath' => null,
+			'valuePath' => null,
+			'spacer' => '_'
+		);
+		$options += $defaults;
 
-		$conditions = (array)$conditions;
-		if ($scope) {
-			$conditions[] = $scope;
+		extract($this->settings[$Model->alias]);
+
+		if (!$options['keyPath']) {
+			$options['keyPath'] = '{n}.' . $Model->alias . '.' . $Model->primaryKey;
```

Comment:
> Because keyPath comes in as null key from generateTreeList().

Don't see why that's an issue. If `keyPath` is `null` we do want to use `'{n}.' . $Model->alias . '.' . $Model->primaryKey` so we can have `'keyPath' => '{n}.' . $Model->alias . '.' . $Model->primaryKey` in `$defaults` itself.